File: algorithms/[69]x 的平方根/solution.ts
// topics = ["数学", "二分查找"]

function mySqrt(x: number): number {
  let left = 0;
  let right = x + 1;

  // 左闭右开写法
  while (left < right) {
    // 这里 mid 的值可能会超过 2^31 - 1， 若使用右移运算符（>>）会有 bug，使用无符号右移代替（>>>）
    const mid = Math.floor((left + right) / 2);
    const square = mid ** 2;

    if (square == x) {
      return mid;
    }

    if (square < x) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  return left - 1;
}

File: algorithms/[69]x 的平方根/solution.py
# topics = ["数学", "二分查找"]


class Solution:
    def mySqrt(self, x: int) -> int:
        left, right = 0, x

        # 闭区间写法
        while left <= right:
            mid = (left + right) // 2
            square = mid ** 2

            if square == x:
                return mid

            if square < x:
                left = mid + 1
            else:
                right = mid - 1

        return left - 1
